wurst/systems/spawns/ResourceSpawnsUtils.wurst

```diff
@@ -0,0 +1,197 @@
+        this.disallowedRegions = disallowedRegions
+    
+    override function execute()
+        let spawnPoint = getRandomAllowedPoint(this.allowedRegions, this.disallowedRegions)
+        let spawnItemId = this.items.random()
+        createItem(spawnItemId, spawnPoint)
+
+        if not this.items.spawnsPerType.has(spawnItemId)
+            this.items.spawnsPerType.put(spawnItemId, 1)
+        else
+            this.items.spawnsPerType.put(spawnItemId, this.items.spawnsPerType.get(spawnItemId) + 1)
+
+public class FishSpawnComposer implements Composer
+    protected Pool<int> units
+    protected rect allowedRegion
+
+    construct(Pool<int> units, rect allowedRegion)
+        this.units = units
+        this.allowedRegion = allowedRegion
+    
+    override function execute()
+        let spawnPoint = this.allowedRegion.randomPoint()
+        let targetID = this.units.random()
+    
+        // Spawn a unit of the selected type.
+        createUnit(
+            players[PLAYER_NEUTRAL_AGGRESSIVE],
+            targetID,
+            spawnPoint,
+            randomAngle()
+        )
+
+public class AnimalSpawnComposer implements Composer
+    protected Pool<int> units
+    protected Pool<rect> allowedRegions
+    protected region disallowedRegions
+
+    construct(Pool<int> units, Pool<rect> allowedRegions, region disallowedRegions)
+        this.allowedRegions = allowedRegions
+        this.disallowedRegions = disallowedRegions
+        this.units = units
+    
+    override function execute()
+        let spawnPoint = getRandomAllowedPoint(this.allowedRegions, this.disallowedRegions)
+        let animal = this.units.random()
+        createUnit(players[PLAYER_NEUTRAL_AGGRESSIVE], animal, spawnPoint, randomAngle())
+
+public class IslandInfo
+    int itemsSpawnCount
+    int animalsSpawnCount
+    Pool<rect> spawnRegionsPool
+    
+    construct(int itemsSpawnCount, int animalsSpawnCount, Pool<rect> spawnRegionsPool)
+        this.itemsSpawnCount = itemsSpawnCount
+        this.animalsSpawnCount = animalsSpawnCount
+        this.spawnRegionsPool = spawnRegionsPool
+
+// From Marsunpaisti :
+// this class handles the spawning system trying to emulate the old code
+// so for example mana crystals at start of game can have like a 5% chance to spawn and it increases to 25% slowly
+public class WeightScaler
+    real initialSpawnWeight
+    real finalSpawnWeight
+    real weightChangeTime
+    construct(real initialSpawnWeight, real finalSpawnWeight, real weightChangeTime)
+        this.initialSpawnWeight = initialSpawnWeight
+        this.finalSpawnWeight = finalSpawnWeight
+        this.weightChangeTime = weightChangeTime
+
+    function getTimeAdjustedSpawnWeight() returns real
+        if this.weightChangeTime == 0
+            return finalSpawnWeight
+        //Clamp ratio between 0 - 1 with min & max
+        let ratio = min(1, max(0, (getElapsedGameTime() - GAME_TIMER.getElapsed()) / this.weightChangeTime))

```

Comment:
I don't understand this line - why is `getElapsedGameTime` considered at all? That would be affected by mode and class selection times.

Reply:
My bad, I used the GAME_TIMER where it wasn't used before, but forgot to remove the -